core/flow/rule_manager.go

```diff
@@ -173,12 +177,92 @@ func buildFlowMap(rules []*FlowRule) TrafficControllerMap {
 		if !exists {
 			m[rule.Resource] = []*TrafficShapingController{tsc}
 		} else {
-			m[rule.Resource] = append(rulesOfRes, tsc)
+			if tsc.rule.ID == 0 || isAppend(rulesOfRes, tsc.rule.ID) {
+				m[rule.Resource] = append(rulesOfRes, tsc)
+			}
 		}
+
 	}
 	return m
 }
 
+func isAppend(rulesOfRes []*TrafficShapingController, ruleId uint64) bool {
+	for _, v := range rulesOfRes {
+		if v.rule.ID == ruleId {
+			logging.Warnf("Rule id:%d duplicate , current rule cannot be loaded", ruleId)
+			return false
+		}
+	}
+	return true
+}
+
+func UpdateRule(rule *FlowRule) error {
+	if err := IsValidFlowRule(rule); err != nil {
+		return err
+	}
+	if rule.ID == 0 {
+		return errors.New("Update rule id cannot be 0")
+	}
+	if rule.LimitOrigin == "" {
+		rule.LimitOrigin = LimitOriginDefault
+	}
+	rulesOfRes, exists := tcMap[rule.Resource]
+
+	if !exists {
+		return errors.New("Cannot be updated ,because not loaded flow rule")
+	}
+	updateFlag := false
+	for k, v := range rulesOfRes {
+		if v.rule.ID == rule.ID {
+			generator, supported := tcGenFuncMap[rule.ControlBehavior]
+			if !supported {
+				return errors.New("Ignoring the rule due to unsupported control behavior")
+			}
+			tsc := generator(rule)
+			if tsc == nil {
+				return errors.New("Ignoring the rule due to bad generated traffic controller")
+			}
+			tcMux.RLock()
+			rulesOfRes[k] = tsc
+			tcMux.RUnlock()
+			updateFlag = true
+		}
+	}
+
+	if !updateFlag {
+		return errors.New(fmt.Sprintf("No rule to update was found based on rule id:%d", rule.ID))
+	}
+	return nil
+}
+
+func AddRule(rule *FlowRule) error {
+	if err := IsValidFlowRule(rule); err != nil {
+		return err
+	}
+	if rule.LimitOrigin == "" {
+		rule.LimitOrigin = LimitOriginDefault
+	}
```

Comment:
Maybe we need to check whether this rule.Id has existed？

Reply:
At appendRule, ruleID is not required; we can check and print warning log notifications.

Reply:
Make sense.
I think logging is not required.
Otherwise,  I think it's better to rename this function to AppendRule, right?